staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/fairqueuing.go

```diff
@@ -0,0 +1,482 @@
+	if timeoutIdx != -1 {
+		// timeoutIdx + 1 to remove the last timeout pkt
+		removeIdx := timeoutIdx + 1
+
+		// remove all the timeout packets
+		queue.SetPackets(pkts[removeIdx:])
+		fqs.DecrementPackets(removeIdx)
+	}
+}
+
+// DecrementPackets decreases the # of packets for the FQScheduler w/ lock
+func (fqs *FQScheduler) DecrementPackets(i int) {
+	fqs.numPackets -= i
+}
+
+// GetRequestsExecuting gets the # of requests which are "executing":
+// this is the# of requests/packets which have been dequeued but have not had
+// finished (via the FinishPacket method invoked after service)
+func (fqs *FQScheduler) getRequestsExecuting() int {
+	total := 0
+	for _, queue := range fqs.Queues {
+		total += queue.GetRequestsExecuting()
+	}
+	return total
+}
+
+//GetQueues returns the queues of the FQScheduler
+func (fqs *FQScheduler) GetQueues() []FQQueue {
+	return fqs.Queues
+}
+
+func (fqs *FQScheduler) lengthOfQueue(i int) int {
+	return len(fqs.Queues[i].GetPackets())
+}
+
+// shuffleDealAndPick uses shuffle sharding to select an index from a set of queues
+func (fqs *FQScheduler) shuffleDealAndPick(v, nq uint64,
+	mr func(int /*in [0, nq-1]*/) int, /*in [0, numQueues-1] and excluding previously determined members of I*/
+	nRem, minLen, bestIdx int) int {
+	if nRem < 1 {
+		return bestIdx
+	}
+	vNext := v / nq
+	ai := int(v - nq*vNext)
+	ii := mr(ai)
+	mrNext := func(a int /*in [0, nq-2]*/) int /*in [0, numQueues-1] and excluding I[0], I[1], ... ii*/ {
+		if a < ai {
+			return mr(a)
+		}
+		return mr(a + 1)
+	}
+	lenI := fqs.lengthOfQueue(ii)
+	if lenI < minLen {
+		minLen = lenI
+		bestIdx = ii
+	}
+	return fqs.shuffleDealAndPick(vNext, nq-1, mrNext, nRem-1, minLen, bestIdx)
+}
+
+// ChooseQueueIdx uses shuffle sharding to select an queue index
+// using a 'hashValue'.  The 'hashValue' derives a hand from a set range of
+// indexes (range 'desiredNumQueues') and returns the queue with the least queued packets
+// from a dealt hand (of size 'handSize')
+func (fqs *FQScheduler) ChooseQueueIdx(hashValue uint64, handSize int) int {
+	// TODO(aaron-prindle) currently a lock is held for this in a larger anonymous function
+	// verify that makes sense...
+
+	// desiredNumQueues is used here instead of numQueues to omit quiesce queues
+	return fqs.shuffleDealAndPick(hashValue, uint64(fqs.desiredNumQueues), func(i int) int { return i }, handSize, math.MaxInt32, -1)
+}
+
+// rejectOrEnqueue rejects or enqueues the newly arrived request if
+// resource criteria isn't met
+func (fqs *FQScheduler) rejectOrEnqueue(packet FQPacket) bool {
+	queue := packet.GetQueue()
+	curQueueLength := len(queue.GetPackets())
+	// rejects the newly arrived request if resource criteria not met
+	if fqs.getRequestsExecuting() >= fqs.concurrencyLimit &&
+		curQueueLength >= fqs.queueLengthLimit {
+		return false
+	}
+
+	fqs.synctime()
+
+	return fqs.enqueue(packet)
+}
+
+// enqueues a packet into an FQScheduler
+func (fqs *FQScheduler) enqueue(packet FQPacket) bool {
```

Comment:
Why is this func declared to return something?  It returns the same value every time.

Reply:
good point, updated to no longer return a value.  The public function `Enqueue` will return a bool but not the private `enqueue` for the reason noted.